tap_nice_incontact/streams.py

```diff
@@ -604,6 +612,188 @@ def get_records(self,
                         for rec in results.get(self.data_key))
 
 
+class DataExtractionStream(IncrementalStream):
+    """
+    Base class for streams pulling data via the Data Extraction API endpoints
+    Docs: https://help.nice-incontact.com/content/recording/dataextractionapi.htm
+    :param client: The API client used extract records from the external source
+    """
+    replication_method = 'INCREMENTAL'
+    entity_name = ''
+
+    # amount of time to delay before polling again
+    poll_delay = 5
+    # number of seconds before timing out
+    poll_timeout = 300
+    default_period = 'days'
+
+    path = 'jobs'
+
+    def get_status_uri(self, job_id):
+        return f'{self.path}/{job_id}'
+
+    def download_csv(self, csv_uri):
+        rows = []
+        with requests.get(csv_uri, stream=True) as r:
+            lines = (line.decode('utf-8') for line in r.iter_lines())
+            for row in csv.DictReader(lines):
+                rows.append(row)
+        return rows
+
+    def create_job(self, start_date, end_date):
+        """ Create a new job in the data extraction API """
+        if DEBUG_JOBS:
+            # Set a fixed job ID for debugging
+            return DEBUG_JOB_ID
+
+        data = {
+            "entityName": self.entity_name,
+            "startDate": start_date.strftime('%Y-%m-%d'),
+            "endDate": end_date.strftime('%Y-%m-%d'),
+            "version": "3",
+        }
+        job_id = self.client.post(
+            self.path,
+            data=json.dumps(data),
+            headers={'Content-Type': 'application/json'},
+            is_data_extraction=True
+        )
+        return job_id
+
+    def get_job_status(self, job_id):
+        job_path = f'{self.path}/{job_id}'
+        results = self.client.get(job_path, is_data_extraction=True)
+        if 'jobStatus' not in results:
+            return None
+        return results['jobStatus']
+
+    def fetch_data_export(self, start_date, end_date):
+
+        job_id = self.create_job(start_date, end_date)
+        
+        LOGGER.info(f'Created job: {job_id}'))
+
+        job_status = None
+        ready = False
+        poll_attempt = 0
+
+        start_time = datetime.datetime.utcnow()
+
+        while not ready:
+            if (datetime.datetime.utcnow() - start_time).total_seconds() > self.poll_timeout:
+                raise Exception("data extraction job status timeout")
+
+            job_status = self.get_job_status(job_id)
+            if job_status is None:
+                raise Exception("data extraction job failure", job_id)
+
+            LOGGER.info(f'Job status ({job_id}): {job_status}')
+
+            if job_status['status'] == 'SUCCEEDED':
+                ready = True
+            elif job_status['status'] == 'RUNNING':
+                time.sleep(self.poll_delay)
+            else:
+                raise Exception("data extraction job failure", job_status)
```

Comment:
In any of these exceptions, do we want to continue to the next start_date and end_date or is it fine to stop since we know that future jobs won't work? Currently calls to this method only catch 403

Reply:
Yep good call, also catching 500s in case the server times out for some reason. Also will throw a `ValueError` instead for invalid job states (Failed, Expired, Cancelled)